Do not import any sbt subprojects in IntelliJ IDEA in which BSP is disabled

vasilmkd opened the pull request:
- This change improves the development experience of the Scala 2 compiler repo in IntelliJ IDEA when using the default sbt project import flow.
- All sbt projects which are not exported to BSP clients are now also not imported into IntelliJ IDEA.

I'd also like to discuss backporting these changes to the 2.12.x branch, which is still actively maintained.

Additionally, I'd like to start a discussion about the current state of the "IDE Setup" [section](https://github.com/scala/scala?tab=readme-ov-file#ide-setup) of the README and potentially contribute some updates to it.

I tried importing the project in IntelliJ IDEA by running the `sbt intellij` script which generates the IDEA project files and opening the repository that way, as recently as last week. I observed that the project could not be built and fails during compilation. On the other hand, just opening the repository directory in IDEA and letting the default sbt import do its thing, fully imports the project and allows it to be compiled without issues.

Thanks in advance.

som-snytt commented on build.sbt:
```diff
@@ -817,11 +826,14 @@ lazy val bench = project.in(file("test") / "benchmarks")
     },
     //scalacOptions ++= Seq("-feature", "-opt:inline:scala/**", "-Wopt"),
-    scalacOptions ++= Seq("-feature", "-opt:l:inline", "-opt-inline-from:scala/**", "-opt-warnings"),
+    scalacOptions ++= Seq("-feature", "-opt:l:inline", "-opt-inline-from:scala/**", "-opt-warnings")
```
the trailing comma was removed. Also the commented option should be used by now. Somehow that change never gets in...

vasilmkd replied:
I reverted the trailing comma.

vasilmkd committed:
Do not import any sbt subprojects in IntelliJ IDEA in which BSP is disabled

- This change improves the development experience of the Scala 2 compiler repo in IntelliJ IDEA when using the default sbt project import flow.
- All sbt projects which are not exported to BSP clients are now also not imported into IntelliJ IDEA.

vasilmkd committed:
Set the dynamically defined "ide-skip-project" key to be Invisible, such that sbt won't print it as unused during initialization

lrytz commented:
Thanks all!

We can remove the template files and `sbt intellij` - I haven't used that in years, sorry for the confusion.

I'm not using sbt import either though. Instead I'm using "Import Project from Existing Sources", then "BSP", then "sbt" (not with Bloop). That works pretty well for me, I can build and run / debug unit tests from the gutter.

I just tested the default sbt import (before this PR), which used to not work in the past (that's why we had `sbt intellij`). That seems to work as well now, so the improvements for that option from this PR are very welcome.

Is there a reason to prefer one over the other?

unkarjedy commented:
@lrytz The default SBT way (non-BSP) is still the default way to go.
The BSP implementation is not actively maintained.
A lot of issues that BSP resolved were fixed in the default/"native" sbt way over the past couple of years.

Note that those things were working in the BSP mode not because the BSP mode had any principle advantages that made it better. They worked just because some time ago the work on the Build Tools support was diverged in two directions and the native approach lacked some attention.

vasilmkd commented:
Additionally, in my opinion, simply opening an sbt project directory in IDEA will use the sbt import, which I would assume is what most users will do (instead of going through `Import Project from existing sources`.

> That works pretty well for me, I can build and run / debug unit tests from the gutter.
For the record, I tested running unit tests in the Scala 2 compiler repo using the regular sbt import. Everything seems to be working.

Could I please ask the maintainers to give this a shot and work with this setup for a few days? Additionally, if you use the just released IntelliJ IDEA 2025.2, there are even more fixes that should benefit this repository specifically.

Thank you in advance.

> We can remove the template files and sbt intellij - I haven't used that in years, sorry for the confusion.
I propose that we merge this current PR as is, and I will work on contributing the other changes in a separate pass, i.e. removing the `sbt intellij` script/documentation, and improving the documentation to point towards the newer, more maintained approaches. WDYT?

lrytz reviewed (approved):
Thank you!

lrytz commented:
@vasilmkd I came across one issue, I cannot do the equivalent of `partest/compile` in IntelliJ, compilation fails for `partest/nest/UnsafeAccess.java` with

```
/Users/luc/code/scala/scala13/src/partest/scala/tools/partest/nest/UnsafeAccess.java:19:33
package sun.misc does not exist
sun.misc.Unsafe
```

The `build.sbt` sets the `-XDenableSunApiLintControl`, and it seems to be picked up by IntelliJ

<img width="1109" height="396" alt="image" src="https://github.com/user-attachments/assets/4d5fae3e-ec3a-43a1-a771-6bcdd6ac8d0b" />

Build output is

```
Searching for compilable files... [scala2.scala-partest.main]
Reading compilation settings... [scala2.scala-partest.main]
java: User-specified option "-g" is ignored for "scala2.scala-partest.main". This compilation parameter is set automatically according to project settings.
Compiling... [scala2.scala-partest.main]
scala: compiling 47 Scala sources and 2 Java sources to /Users/luc/code/scala/scala13/build/quick/classes/partest ...
Errors occurred while compiling module 'scala2.scala-partest.main'
Finished, saving caches…
```

I think the error message comes from the Java compiler. Could you take a look..?

lrytz commented:
Nice, thanks! Same for me: it works with Java 8, and it works with Java 21 when disabling the setting you mention.
